### ISSUE TITLE
Improve typing in plan stubs, generally and to allow using Ophyd devices

### ISSUE THREAD

**Issue by d-perl** [bluesky/bluesky#1809]
Recent additions of type hints have some errors and omissions - importantly they aren't compatible with function signatures and devices from Ophyd 

## Expected Behavior

Scanning with `pyright` on code which uses bluesky library functions correctly applied to ophyd devices should not raise errors

## Current Behavior
Something like
```python
bps.mv(backlight, BacklightPosition.OUT)
```
will give
```
Argument of type "Backlight" cannot be assigned to parameter "args" of type "Tuple[Movable | NamedMovable | Any, ...]" in function "mv"
    "Backlight" is not assignable to "Tuple[Movable | NamedMovable | Any, ...]" (reportArgumentType)
```

or something like 

```python
bps.abs_set(motor, position)
```
will give 
```
 error: Argument of type "EpicsMotor | Motor" cannot be assigned to parameter "obj" of type "Movable" in function "abs_set"
    Type "EpicsMotor | Motor" is not assignable to type "Movable"
      "EpicsMotor" is incompatible with protocol "Movable"
        "set" is an incompatible type
          Type "(new_position: Any, *, timeout: float = None, moved_cb: (...) -> Unknown = None, wait: bool = False) -> StatusBase" is not assignable to type "(value: Unknown) -> Status"
            Parameter name mismatch: "value" versus "new_position"
            "None" is not assignable to "float"
            Type "None" is not assignable to type "(...) -> Unknown"
            Function return type "StatusBase" is incompatible with type "Status" (reportArgumentType)

```
## Possible Solution
- Expand the signatures for the plan stubs to accommodate Ophyd devices and not only ophyd-async
- Fix or remove errors like the above `tuple` definition

## Your Environment
Encountered in dodal CI runs

**Comment by dperl-dls** [comment]
We (@callumforrester @DiamondJoseph ) think that the hint for `*args` in `plan_stubs.mv` can just be removed since it reduces to `tuple[Any]` anyway, which is anything that `*args` could possibly ever be, and there's not currently any way to specify such a repeating pattern of arguments as `device1, value1, device2, value2, ...`.

```python
@plan
def mv(
    *args: Tuple[Union[Movable, NamedMovable, Any], ...],
    group: Optional[Hashable] = None,
    **kwargs,
) -> MsgGenerator[Tuple[Status, ...]]:
```

**Comment by dperl-dls** [comment]
I believe that `trigger`

```python
@plan
def trigger(
    obj: Triggerable,
    *,
    group: Optional[Hashable] = None,
    wait: bool = False,
) -> MsgGenerator[Status]:
```

Should expect a `StatusBase` instead of `Status` as it currently doesn't work with `DeviceStatus` which only inherits from the former, and only requires the status to be `wait`-able

**Comment by DiamondJoseph** [comment]
> We think that the hint for `*args` in `plan_stubs.mv` can just be removed...

We might actually be able to make use of [typing.Unpack](https://docs.python.org/3.12/library/typing.html#typing.Unpack)

```python
from typing_extensions import TypedDict, Unpack # < Python 3.12

@plan
def mv(
    *args: Unpack[Tuple[Union[Movable, NamedMovable], Any]],
    group: Optional[Hashable] = None,
    **kwargs,
) -> MsgGenerator[Tuple[Status, ...]]:
```

**Comment by dperl-dls** [comment]
yeah but doesn't that only work for the case of one device and one value? I still think we can't get a repeating version?

**Comment by DiamondJoseph** [comment]
> yeah but doesn't that only work for the case of one device and one value? I still think we can't get a repeating version?

Does it not? `*args` should be consistent for all N>=0

**Comment by dperl-dls** [comment]
As I understand it, the whole thing about `Unpack` is to do the thing that the original hint is trying to do with `tuple` - and the above is actually interpreted as "every member of `*args` must be a `tuple[Movable, Any, ...]`"

**Comment by dperl-dls** [comment]
Ophyd devices also don't conform to `Readable`, so `bps.read()` complains that "read" is an incompatible type

```
      Type "() -> OrderedDictType" is not assignable to type "() -> SyncOrAsync[Dict[str, Reading[Unknown]]]"
        Function return type "OrderedDictType" is incompatible with type "SyncOrAsync[Dict[str, Reading[Unknown]]]"
          Type "OrderedDictType" is not assignable to type "SyncOrAsync[Dict[str, Reading[Unknown]]]"
    "describe" is an incompatible type
      Type "() -> OrderedDictType" is not assignable to type "() -> SyncOrAsync[Dict[str, DataKey]]"
        Function return type "OrderedDictType" is incompatible with type "SyncOrAsync[Dict[str, DataKey]]"
```

**Comment by dperl-dls** [comment]
@DominicOram suggested it would be nice to be able to use any argument name for `set` in `Movable`, the following demonstrates that 

```python
    def set(self, *args: Unpack[Tuple[Any]]) -> StatusBase:
```

should work for that

```python
from abc import abstractmethod
from typing import Any, Protocol, SupportsFloat, Tuple, Unpack, runtime_checkable


@runtime_checkable
class TakesOnlyValue(Protocol):
    @abstractmethod
    def set(self, value) -> SupportsFloat:
        """Take an arg called value and return a float"""
        ...

@runtime_checkable
class TakesAnySingleArgName(Protocol):
    @abstractmethod
    def set(self, *args: Unpack[Tuple[Any]]) -> SupportsFloat:
        """Take any arg and return a float"""
        ...

class GoodTakesOnlyValue:
    def set(self, value):
        return 1

class BadTakesOnlyValue:
    def set(self, not_value):
        return 1

class GoodTakesAnyArg1:
    def set(self, position):
        return 2.0

class GoodTakesAnyArg2:
    def set(self, value):
        return 2.0

class GoodTakesAnyArg3:
    def set(self, setting):
        return 2.0

class BadTakesAnyArg:
    def set(self, setting, other):
        return 2.0

def set_any_arg_name(obj: TakesAnySingleArgName):
    obj.set(1)

def set_only_value(obj: TakesOnlyValue):
    obj.set(1)


set_any_arg_name(GoodTakesAnyArg1())
set_any_arg_name(GoodTakesAnyArg2())
set_any_arg_name(GoodTakesAnyArg3())
set_any_arg_name(BadTakesAnyArg()) # pyright complains

set_only_value(GoodTakesOnlyValue())
set_only_value(BadTakesOnlyValue()) # pyright complains
set_only_value(GoodTakesAnyArg3()) # pyright complains

assert(isinstance(GoodTakesAnyArg2(), TakesAnySingleArgName))
assert(not isinstance(BadTakesAnyArg(), TakesAnySingleArgName))
assert(isinstance(GoodTakesOnlyValue(), TakesOnlyValue))
assert(not isinstance(GoodTakesOnlyValue(), TakesOnlyValue))

```

**Comment by DominicOram** [comment]
> @DominicOram suggested it would be nice to be able to use any argument name for `set` in `Movable`, the following demonstrates that

My argument here is that what is being `set` on a device will vary wildly across devices. If, when writing a device, you need to use `value` you will invariably have to then have some comments/docstrings to explain what `value` is in this context. It would be more self documenting to let the device use something context specific.

I could be convinced that this is a bad idea if it's shown that something, currently or in the future, depends on this being a `value` e.g. using it as a keyword arg.

**Comment by dperl-dls** [comment]
I believe that at least `trigger`, and many of the other plan stubs

```python
@plan
def trigger(
    obj: Triggerable,
    *,
    group: Optional[Hashable] = None,
    wait: bool = False,
) -> MsgGenerator[Status]:
```

Should expect a `StatusBase` instead of `Status` as it currently doesn't work with `DeviceStatus` which only inherits from the former, and only requires the status to be `wait`-able

**Comment by olliesilvester** [comment]
Also, there are a few plan stubs that don't have anything typed as a return value (I think they should all have `MsgGenerator`). Eg `wait` and `prepare`. This causes pyright to complain if you use `bps.wait` in a function that accepts a `MsgGenerator`

**Comment by coretl** [comment]
@DiamondJoseph @dperl-dls @callumforrester are any of you working on a fix for these issues? If no quick fix is imminent should we widen the signatures of `bps.mv` and `bps.rd` (and anything else reported in this thread) to `Any` so we can avoid peppering our code bases with `# type: ignore`s?

**Comment by DiamondJoseph** [comment]
Given we're going to be wrapping a bunch of bluesky.plans to get our required filesystem handling, and we previously had wrapped versions of the stubs for -> MsgGenerator, we could continue to have wrapped versions of all of the plans/stubs in perpetuity: I just worry that at some point we'll end up with a mix that will never be unmixed. I'm not actively working on it at the moment.

**Comment by tacaswell** [comment]
> Should expect a StatusBase instead of Status as it currently doesn't work with DeviceStatus which only inherits from the former, and only requires the status to be wait-able

I agree that seems correct.  I suspect this is fuzzier docs for humans being translated too directly to types for the computer.

> 
Ophyd devices also don't conform to Readable, so bps.read() complains that "read" is an incompatible type

That is unfortunate, but I don't quite understand why we we can not return an ordered dict instead of a dict?  I suspect the problem is the ophyd typing is lacking the `[str, Reading[Unknown]]` specification?  We stated back before dict iteration order was stable (the "it is random every time" era was fun) so we returned `OrderedDict`s to cope with that.  However for back-compatibility we kept returning `OrderedDicts` in case someone was relying on the extra API they carry.  I really hope there is a way to make the typing more verbose to fix this rather than having to purge `OrderedDict` from ophyd.

**Comment by coretl** [comment]
> ```python
> @plan
> def trigger(
>     obj: Triggerable,
>     *,
>     group: Optional[Hashable] = None,
>     wait: bool = False,
> ) -> MsgGenerator[Status]:
> ```

It looks like the plan stubs import from [`bluesky.protocols`](https://github.com/bluesky/bluesky/blob/3bce96b85558993499174be7eebcd09f270a9271/src/bluesky/plan_stubs.py#L49), so this is not an ophyd `Status`, this is a bluesky protocol `Status`. Both ophyd's `Status` and `StatusBase` (so hopefully also `DeviceBase`) conform to `bluesky.protocol.Status`, so I don't understand why pyright is complaining? Can you post the error @dperl-dls ?

**Comment by dperl-dls** [comment]
@DiamondJoseph I just started looking at actually fixing this. In the first instance I will just widen the signatures to accept ophyd things / be correct. If there is anything which actually requires changes to what happens at runtime I will make a separate issue for that, I don't imagine it should be necessary.

**Comment by dperl-dls** [comment]
> That is unfortunate, but I don't quite understand why we we can not return an ordered dict instead of a dict? I suspect the problem is the ophyd typing is lacking the [str, Reading[Unknown]] specification? We stated back before dict iteration order was stable (the "it is random every time" era was fun) so we returned OrderedDicts to cope with that. 

The problem is not that it wants a non-ordered dict but that `Readable` protocol expects the diccts of the specific `TypedDict`s `Reading[T]` and `DataKey` (from `event_model`) to be returned from `Device.read()` and `describe()` respectively. I think that could be fixed in ophyd's `device.py`'s `BlueskyInterface`, `PositionerBase`, but would require importing DataKey from `event_model` and `Reading` from `bluesky.protocols`, which I guess ophyd doesn't want to depend on?

Given the complications of the above, I'm going to first just fix this for `ophyd-async` and then we can further discuss what to do about ophyd, given I think the situation where people are using strict type checking for ophyd is relatively rare (at least, nobody seems to have reported any problems here)

**Comment by dperl-dls** [comment]
@coretl as for the status issue, it was actually that the ophyd `Status`es don't conform to `bluesky.protocols.Status`

```
rgument of type "StatusBase" cannot be assigned to parameter "s" of type "Status" in function "takes_status"
  "StatusBase" is incompatible with protocol "Status"
    "exception" is an incompatible type
      Type "(timeout: Unknown | None = None) -> (StatusTimeoutError | Exception | type[Exception] | None)" is not assignable to type "(timeout: float | None = 0) -> (BaseException | None)"
        Function return type "StatusTimeoutError | Exception | type[Exception] | None" is incompatible with type "BaseException | None"
          Type "StatusTimeoutError | Exception | type[Exception] | None" is not assignable to type "BaseException | None"
    "Status" is not assignable to "StatusBase"
```

**Comment by olliesilvester** [comment]
I believe #1871 fixed the typing for `Movable`, but not for `Readable`, following @d-perl's comment from october:

> Ophyd devices also don't conform to `Readable`, so `bps.read()` complains that "read" is an incompatible type
> 
> ```
>       Type "() -> OrderedDictType" is not assignable to type "() -> SyncOrAsync[Dict[str, Reading[Unknown]]]"
>         Function return type "OrderedDictType" is incompatible with type "SyncOrAsync[Dict[str, Reading[Unknown]]]"
>           Type "OrderedDictType" is not assignable to type "SyncOrAsync[Dict[str, Reading[Unknown]]]"
>     "describe" is an incompatible type
>       Type "() -> OrderedDictType" is not assignable to type "() -> SyncOrAsync[Dict[str, DataKey]]"
>         Function return type "OrderedDictType" is incompatible with type "SyncOrAsync[Dict[str, DataKey]]"
> ```

**Comment by evalott100** [comment]
>  fixed the typing for Movable, but not for Readable

Will fix now

**Comment by evalott100** [comment]
Closing since https://github.com/bluesky/bluesky/issues/1809#issuecomment-2630708730 should be handled in ophyd:

https://github.com/bluesky/ophyd/issues/1232
